### ISSUE TITLE
Post processing speed and geojson tile alignment

### ISSUE THREAD

**Issue by LukasOldenburg** [digitalpathologybern/hover_next_inference#8]
Hi,

thanks for this amazing work and impact on H&E phenotyping. 

I did inference on a .mrxs WSI at 40x resolution with the following settings:

params = {
    "input": "/path/to/aaa.mrxs",
    "output_root": "./results",
    "cp": "pannuke_convnextv2_tiny_1",
    "only_inference": False,
    "metric": "f1",
    "batch_size": 256,
    "tta": 4,
    "save_polygon": True,
    "tile_size": 256,
    "overlap": 0.9375,
    "inf_workers": 16,
    "inf_writers": 4,
    "pp_tiling": 8,
    "pp_overlap": 256,
    "pp_workers": 16,
    "keep_raw": False,
    "cache": None,
}
main(params)
_____________________________________________________________
OUTPUT LOG 
processing input using pannuke trained model
Running inference on 1 file(s)
100%| 193/193 [04:17<00:00,  1.33s/it]
::: finished or skipped inference after 0:04:18.413159
running post-processing
[0.7 0.8 0.4 0.2 0.7] [0.5 0.6 0.5 0.4 0.6]
getting coords:
100%| ... 64/64 [10:45:07<00:00, 604.81s/it]
::: postprocessing took 10:52:57.480549 total elapsed time 10:52:57.490949
______________________________________________________________

1) Very cool to see that you can do processing and inference in roughly 5 minutes on a WSI in this specific case. However, I was wondering why the postprocessing takes so much longer (~10 hours) and if there are potentially ways to speed this up.

2) When I load the poly.geojson into QuPath, I see a misalignment of the tiles. They are at the correct position in the WSI but the cell masks seem to have a wrong scaling (roughly factor x2 missing, see screenshot). 
![Screenshot from 2024-07-17 11-28-52](https://github.com/user-attachments/assets/6c05102f-77ae-4516-ac72-c7c8ca0cdeff)

Thank you and happy to hear back from you,
Lukas

**Comment by eliasbaumann** [comment]
Hey Lukas,

thanks for opening this issue! Do you mind sharing the specs of the machine you are running this on? And are you loading data from a network drive or is it stored on the machine itself? Given the time it took for inference, the postprocessing time is really excessive and should be more in the 5-10min range, but the post-processing is reading large zarr arrays so if that takes a while then that might be an explanation?
As a starting point, you could try with fewer `pp_workers` and see if that changes anything, and check if you can put the output directory to a local scratch storage in case you are reading/writing from and to the network?

For your second question: 
I am not really using the geojson output myself, but I quickly tried with an mrxs slide and dont have the issue, so we might have to dig a little deeper to figure out why this is happening. It seems though that it has to be an issue with the post-processing already, so maybe thats also part of why it takes so long. The geojson output simply takes the instance map and json class map to produce an output so the same issue should be in the actual raw array. Did you modify anything in the parameter file for the actual model (under /pannuke_convnextv2_tiny_1/params.toml)?

Can you run this and post the output:

```
import openslide
sl = openslide.open_slide("/path/to/aaa.mrxs")
print(sl.properties)
```

**Comment by LukasOldenburg** [comment]
Hi Elias,

thanks for the quick answer, appreciate it. I'm running a Lambda worstation locally, with a 32-core AMD Ryzen Threadripper PRO 5955WX, 128GB RAM. Also having a NVIDIA GeForce RTX 4090. All the data is stored on a fast SSD, so data transfer I/O shouldn't be a bottleneck. Lowering the pp_workers to 4 does not change much. However, I noticed that the post-processing is relatively fast until 12% of the tiles are processed and then slows down very abrupt for the rest of the tiles. 

Some hints: The geojson function seems to work fine, the generation of the pinst_pp.zip seems to have an issue scaling the masks correctly (see screenshot). 

![Screenshot from 2024-07-29 13-29-54](https://github.com/user-attachments/assets/d2882699-0e4f-4873-b625-fc9f80302598)

Also wenn I use the lizard model, the masks are scaled by factor 3x (see other screenshot) wrong compared to pannuke where it's roughly 2x.

![Screenshot from 2024-07-29 13-29-03](https://github.com/user-attachments/assets/8b17bb15-c74c-4d90-96e4-101e1167f405)

**Comment by eliasbaumann** [comment]
Ah, there is definitely an issue with your image being scanned at super high res which is what is causing the stitching issue. I have some images with the same resolution so will work on a fix so that this will not happen anymore and push that. ('openslide.mpp-x': '0.121398698884758').

About the abrupt slowdown at 12% I have no clue so far, but I am thinking its because those first tiles are mostly empty so watershed will be fast. The two problems might be linked together actually, since the output file is also way to large. I'll run an example at high res and see if that also happens for me and investigate then. Thanks for your efforts!

**Comment by eliasbaumann** [comment]
Okay, I maybe fixed it. Can you pull from the branch [faster-geojson ](https://github.com/digitalpathologybern/hover_next_inference/tree/faster-geojson) and see if that

a) fixes your speed problem and 

b) fixes that output issues? 

I ran a 0.12mpp mirax (.mrxs) WSI and it worked fine for me, both inference and postprocessing at reasonable speeds.

**Comment by LukasOldenburg** [comment]
Runs perfectly smooth now, ~10 mins for inference+postprocessing. The outputs are correct as well.
Just out of interest: Can you comment on the issue? What was the problem?

Thank you,
Lukas

**Comment by eliasbaumann** [comment]
The dataloader for inference already accounts for 80x/0.125mpp, but the post-processing, in particular for the tile stitching, did not. There I basically hardcoded the scaling factor, but now its estimated automatically (and will raise an error in case you are trying to run the model on an image with a too low resolution).

Post-processing then ran on these rather big ROIs with the empty space between tiles. the function faster_instance_seg divides the ROIs further into connected areas of predictions, but here there were not connected predictions since everything was at least one tile apart. So therefore every tile was essentially processed individually and sequentially.

Thanks for taking the time to help me figure this out 👍
